typescript/connections.md

```diff
@@ -0,0 +1,151 @@
+
+```ts
+// worker.ts
+await Core.install({ serverOptions: { address, namespace } });
+// Worker uses a connection created by the Core singleton
+const worker = await Worker.create({ taskQueue });
+```
+
+## Proposals
+
+### Explicit `Connection` objects
+
+- `Connection` in `@temporalio/client` for client only
+- `NativeConnection` in `@temporalio/worker` for sharing connections between different workers
+
+```ts
+// worker.ts
+import { NativeConnection, Worker } from '@temporalio/worker';
+
+const connection = await NativeConnection.create({ address });
+const worker = await Worker.create({ connection, namespace, taskQueue });
+
+// starter.ts
+
+// same as today but Connection becomes async to make the `getSystemInfo` call
+import { Connection, WorkflowClient } from '@temporalio/client';
+
+const connection = await Connection.create({ address });
+const client = new WorkflowClient(connection.service, { namespace });
+```
+
+#### Pros
+
+- Explicit connection reuse
+
+#### Cons
+
+- Verbose
+- Confusing when to use each connection type
+
+### Implicit connection reuse with similar options
+
+```ts
+// worker.ts
+import { Worker } from '@temporalio/worker';
+
+// worker1 and worker2 share the same underlying connection based on given options
+const worker1 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq1' });
+const worker2 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq2' });
+
+// starter.ts
+import { WorkflowClient } from '@temporalio/client';
+
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+```
+
+#### Pros
+
+- Short
+- Easy path from sample with local server to real world use case
+
+#### Cons
+
+- Implicit connection sharing is non-obvious
+- No way to opt out of connection sharing (can be addressed if needed)
+
+### Mix of options and explicit clients (recommended)
+
+Support a variant of both approaches listed above with no implicit connection reuse
+
+```ts
+// worker.ts
+import { NativeConnection, Worker } from '@temporalio/worker';
+
+// worker1 and worker2 do not share the same connection
+const worker1 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq1' });
+const worker2 = await Worker.create({ connectionOptions: { address }, namespace, taskQueue: 'tq2' });
+
+// worker3 and worker4 share the same connection
+const connection = await NativeConnection.create({ address });
+const worker3 = await Worker.create({ connection, namespace, taskQueue });
+const worker4 = await Worker.create({ connection, namespace, taskQueue });
+
+// starter.ts
+import { Connection, WorkflowClient } from '@temporalio/client';
+
+const client = await WorkflowClient.create({ connectionOptions: { address }, namespace });
+
+const connection = await Connection.create({ address });
+const client = new WorkflowClient({ connection, namespace });
+```
+
+### Python SDK approach
```

Comment:
FWIW, this is also the Go and Java SDK approach. This is more accurately called the "Every SDK we have except TypeScript approach".

Reply:
It's not the Java approach because Java has `WorkflowClient` and the service stub.

Reply:
Yes, but that'd be more like the explicit connection proposal not passing around the high level client.